In [1]:
import numpy as np
import pandas as pd

In [2]:
from rpy2.robjects import r
from rpy2.robjects.pandas2ri import ri2py

r('library(MASS)')
boston = ri2py(r('Boston'))


We add few columns and datats for exercise proposes 

In [3]:
boston['inverse_zn'] = 1/boston['zn']
boston['ones'] = boston['inverse_zn'] * boston['zn']
boston['stupid'] = 'stupid'

boston2 = boston[['medv', 'ones']]

boston_lstat = boston[['lstat', 'medv']]
boston.head()

crim    zn  indus  chas    nox     rm   age     dis  rad    tax  \
1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   ptratio   black  lstat  medv  inverse_zn  ones  stupid  
1     15.3  396.90   4.98  24.0    0.055556   1.0  stupid  
2     17.8  396.90   9.14  21.6         inf   NaN  stupid  
3     17.8  392.83   4.03  34.7         inf   NaN  stupid  
4     18.7  394.63   2.94  33.4         inf   NaN  stupid  
5     18.7  396.90   5.33  36.2         inf   NaN  stupid

### 1. Renaming columns in pandas
http://stackoverflow.com/questions/11346283/renaming-columns-in-pandas

Rename columns in boston_lstat to 'lower_status', 'median_value'

In [4]:
boston_lstat.columns = ['lower_status', 'median_value']

### 2. Delete column from pandas DataFrame
http://stackoverflow.com/questions/13411544/delete-column-from-pandas-dataframe

Delete 'stupid' column from boston.

In [5]:
if 'stupid' in boston.columns:
    del boston['stupid']
boston.columns

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'black', 'lstat', 'medv', 'inverse_zn', 'ones'],
      dtype='object')

### 3. Adding new column to existing DataFrame in Python pandas

http://stackoverflow.com/questions/12555323/adding-new-column-to-existing-dataframe-in-python-pandas

Create boston3 DataFrame that contains all rows from boston2 that have 'medv' > 24. Then create new column 'normal' and assign to each row a number sampled from normal distribution with mu=100 and sigma=20.

In [8]:
# Create boston3
boston3 = boston2.loc[boston2['medv'] > 24].copy()
boston3.head()

medv  ones
3  34.7   NaN
4  33.4   NaN
5  36.2   NaN
6  28.7   NaN
8  27.1   1.0

In [12]:
# create the array with the normal distribution values
import numpy as np

mean = 3
sd = 0.5
n = boston3.shape[0]

sampled = np.random.normal(loc=mean, scale=sd, size=n)

In [13]:
boston3['normal'] = sampled

### 4. add one row in a pandas.DataFrame

http://stackoverflow.com/questions/10715965/add-one-row-in-a-pandas-dataframe

I understand that pandas is designed to load fully populated DataFrame but I need to create an empty DataFrame then add rows, one by one. What is the best way to do this ?

Create empty DataFrame 'best_dses' with columns 'people' and 'salary', and then add rows, one by one:
"Carlos" 1,000,000, 
"Aleix" 10,000
"Bartek" 8,000. 

In [21]:
best_dses = pd.DataFrame(columns = ['name','salary'])
best_dses.head()

best_dses.loc[0] = ['Carlos', 1000000]
best_dses.loc[1] = ['Aleix', 10000]
best_dses.loc[2] = ['Bartek', 20000]

best_dses

name     salary
0  Carlos  1000000.0
1   Aleix    10000.0
2  Bartek    20000.0

### 5. How to drop rows of Pandas dataframe whose value of certain column is NaN
http://stackoverflow.com/questions/13413590/how-to-drop-rows-of-pandas-dataframe-whose-value-of-certain-column-is-nan

Do it with boston2's column 'ones'.

In [24]:
boston2.dropna(axis = 0, subset=['ones'])

medv  ones
1    24.0   1.0
7    22.9   1.0
8    27.1   1.0
9    16.5   1.0
10   18.9   1.0
11   15.0   1.0
12   18.9   1.0
13   21.7   1.0
40   30.8   1.0
41   34.9   1.0
51   19.7   1.0
52   20.5   1.0
53   25.0   1.0
54   23.4   1.0
55   18.9   1.0
56   35.4   1.0
57   24.7   1.0
58   31.6   1.0
59   23.3   1.0
60   19.6   1.0
61   18.7   1.0
62   16.0   1.0
63   22.2   1.0
64   25.0   1.0
65   33.0   1.0
66   23.5   1.0
67   19.4   1.0
68   22.0   1.0
69   17.4   1.0
70   20.9   1.0
..    ...   ...
288  23.2   1.0
289  22.3   1.0
290  24.8   1.0
291  28.5   1.0
292  37.3   1.0
293  27.9   1.0
299  22.5   1.0
300  29.0   1.0
301  24.8   1.0
302  22.0   1.0
303  26.4   1.0
304  33.1   1.0
305  36.1   1.0
306  28.4   1.0
307  33.4   1.0
308  28.2   1.0
332  17.1   1.0
333  19.4   1.0
342  32.7   1.0
344  23.9   1.0
345  31.2   1.0
348  23.1   1.0
349  24.5   1.0
350  26.6   1.0
351  22.9   1.0
352  24.1   1.0
353  18.6   1.0
354  30.1   1.0
355  18.2   1.0
356  20.6   1.0

[134 rows x 2 columns]

### 6. Pandas: change data type of columns
http://stackoverflow.com/questions/15891038/pandas-change-data-type-of-columns

Change type of column salary in best_dses to int (or to float if already it is int). 

### 7. Selecting columns
http://stackoverflow.com/questions/11285613/selecting-columns

Select columns 'age' and 'lstat' from boston and save it to boston_select DataFrame

### 8. Get list from pandas DataFrame column headers
http://stackoverflow.com/questions/19482970/get-list-from-pandas-dataframe-column-headers

get list of column names of boston

### 9. How to filter the DataFrame rows of pandas by “within”/“in”? (use a list of values to select rows from a pandas dataframe)

http://stackoverflow.com/questions/12065885/how-to-filter-the-dataframe-rows-of-pandas-by-within-in
http://stackoverflow.com/questions/12096252/use-a-list-of-values-to-select-rows-from-a-pandas-dataframe

Select row that has values from list [24, 21.6]

### 10. Convert list of dictionaries to Dataframe
http://stackoverflow.com/questions/20638006/convert-list-of-dictionaries-to-dataframe

Transform best_dses.to_dict('records') to DataFrame